File: predecessors/scraper-backend-python/src/payments/subscription/paystack.py
import requests

class PaystackSubscriptionManager:
    def __init__(self, secret_key):
        self.secret_key = secret_key
        self.base_url = 'https://api.paystack.co'

    def create_subscription(self, customer_id, plan_id):
        url = f'{self.base_url}/subscription'
        headers = {
            'Authorization': f'Bearer {self.secret_key}',
            'Content-Type': 'application/json'
        }
        data = {
            'customer': customer_id,
            'plan': plan_id
        }

        response = requests.post(url, headers=headers, json=data)
        if response.status_code == 201:
            subscription_data = response.json()
            return subscription_data['data']['subscription_code']
        else:
            return None

    def update_subscription(self, subscription_code, customer_id=None, plan_id=None):
        url = f'{self.base_url}/subscription/{subscription_code}'
        headers = {
            'Authorization': f'Bearer {self.secret_key}',
            'Content-Type': 'application/json'
        }
        data = {}
        if customer_id:
            data['customer'] = customer_id
        if plan_id:
            data['plan'] = plan_id

        response = requests.put(url, headers=headers, json=data)
        return response.status_code == 200

    def delete_subscription(self, subscription_code):
        url = f'{self.base_url}/subscription/{subscription_code}'
        headers = {
            'Authorization': f'Bearer {self.secret_key}'
        }

        response = requests.delete(url, headers=headers)
        return response.status_code == 204

File: predecessors/scraper-backend-go/internal/payments/subscriptions/paystack.go
package subscriptions

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

type Subscription struct {
	ID       string `json:"id"`
	Customer string `json:"customer"`
	Plan     string `json:"plan"`
}

type PaystackSubscriptionManager struct {
	secretKey string
}

func NewPaystackSubscriptionManager(secretKey string) *PaystackSubscriptionManager {
	return &PaystackSubscriptionManager{
		secretKey: secretKey,
	}
}

func (manager *PaystackSubscriptionManager) CreateSubscription(s Subscription) (*Subscription, error) {
	url := "https://api.paystack.co/subscription"
	method := "POST"

	payload, err := json.Marshal(s)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest(method, url, strings.NewReader(string(payload)))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", "Bearer "+manager.secretKey)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var subscription Subscription
	err = json.NewDecoder(resp.Body).Decode(&subscription)
	if err != nil {
		return nil, err
	}

	return &subscription, nil
}

func (manager *PaystackSubscriptionManager) GetSubscription(subscriptionID string) (*Subscription, error) {
	url := fmt.Sprintf("https://api.paystack.co/subscription/%s", subscriptionID)
	method := "GET"

	req, err := http.NewRequest(method, url, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", "Bearer "+manager.secretKey)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var subscription Subscription
	err = json.NewDecoder(resp.Body).Decode(&subscription)
	if err != nil {
		return nil, err
	}

	return &subscription, nil
}

func (manager *PaystackSubscriptionManager) UpdateSubscription(subscriptionID string, s Subscription) error {
	url := fmt.Sprintf("https://api.paystack.co/subscription/%s", subscriptionID)
	method := "PUT"

	payload, err := json.Marshal(s)
	if err != nil {
		return err
	}

	req, err := http.NewRequest(method, url, strings.NewReader(string(payload)))
	if err != nil {
		return err
	}

	req.Header.Set("Authorization", "Bearer "+manager.secretKey)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	return nil
}

func (manager *PaystackSubscriptionManager) DeleteSubscription(subscriptionID string) error {
	url := fmt.Sprintf("https://api.paystack.co/subscription/%s", subscriptionID)
	method := "DELETE"

	req, err := http.NewRequest(method, url, nil)
	if err != nil {
		return err
	}

	req.Header.Set("Authorization", "Bearer "+manager.secretKey)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	return nil
}
